### DISCUSSION TITLE
Webots simulation with app layer

### DISCUSSION THREAD

**Discussion by williamleong** [bitcraze/discussions#1000] [Q&A] [answered]
Is it possible to simulate and run code in the app layer such as @knmcguire's SGBA with the python bindings mentioned in https://github.com/orgs/bitcraze/discussions/796#discussioncomment-6610186?

**Comment by knmcguire** [comment, marked as answer]
Ah no not yet... we don't have python binding possibilities for app layer yet.

You could for instance, write a wrapper around the crucial part of the c code and make a binding for that? I've done that in an older branch of last year: https://github.com/bitcraze/crazyflie-simulation/blob/wallfollowing2/webots/controllers/crazyflie_controller_py/crazyflie_controller_py.py

This had an exact copy of the [wall following code of the crazyflie firmwar](https://github.com/bitcraze/crazyflie-firmware/blob/master/examples/demos/app_wall_following_demo/src/wall_following.c)e at the time which [I copied to the repo](https://github.com/bitcraze/crazyflie-simulation/tree/wallfollowing2/controllers) and made a python binding with swig. It might give some inspiration on how to fix that.  The wall_follower.c code was written with that intend in mind that it is very seperate from any platform specific code and only contained logic.

**Reply by williamleong** [reply]
That's a possible option, thanks!

**Reply by knmcguire** [reply]
Ah good! let us know if it works out. If you have any more questions you can start a new discussion 😄
